trace/trace.go

```diff
@@ -28,29 +28,46 @@ import (
 type traceStep struct {
 	stepTime time.Time
 	msg      string
+	msgArgs  []interface{}
 }
 
 // Trace keeps track of a set of "steps" and allows us to log a specific
 // step if it took longer than its share of the total allowed time
 type Trace struct {
-	name      string
-	startTime time.Time
-	steps     []traceStep
+	name        string
+	details     string
+	detailsArgs []interface{}
+	startTime   time.Time
+	steps       []traceStep
 }
 
-// New creates a Trace with the specified name
+// New creates a Trace with the specified name. The name identifies the operation to be traced.
 func New(name string) *Trace {
-	return &Trace{name, time.Now(), nil}
+	return &Trace{name: name, startTime: time.Now()}
+}
+
+// WithDetailsf adds a formatted details string to the trace. The details should include
+// useful data about the trace, such as operation inputs.
+func (t *Trace) WithDetailsf(details string, args ...interface{}) *Trace {
+	t.details = details
+	t.detailsArgs = args
+	return t
 }
 
 // Step adds a new step with a specific message. Call this at the end of an
 // execution step to record how long it took.
 func (t *Trace) Step(msg string) {
+	t.Stepf(msg)
```

Comment:
existing calls here that added step messages containing `%` chars should continue working as-is

Reply:
Exactly my concern on reading this.  You will have to escape all `%` to `%%` in order to do this, or carry a flag all the way through the lifetime that says whether to do format-string expansion later.

Reply:
Thanks for catching this. I've fixed and added a test to ensure Step() continues to work exactly as before with literal strings.